rails_programming/Rails_Sprinkles/JS_Bundling.md

```diff
@@ -0,0 +1,177 @@
+~~~
+
+This command  starts by calling yarn with the build command which processes all of the entry points to your JavaScript into  ```app/assets/builds/application.js```. the ```--watch``` flag allows the bundler to watch for changes to JavaScript files within your project. When these changes happen it's starts the process of bundling those entry points into your JavaScript and updating your project. When doing this you need to be running ```rails server``` in another terminal as well. You will notice when you enter the command that the program will still be running and say ```[watch] build finished, watching for changes...```
+that is the intended behavior. Then it will wait for changes to occur in your JavaScript files and proceed to rebundle them which will update the bundled file that is delivered to the asset pipeline. Another useful command you can use is ```./bin/dev``` which will start the ```yarn build --watch``` and ```rails server``` at the same time. <span id="yarn-build-command"></span>Meaning no having to enter to switch between terminal windows for developing. In development mode the command ```./bin/dev``` is useful since it allows your javascript bundle watcher and rails server to update without needing to manually update it as yarn watches for those changes automatically.</span>
+
+
+
+### Using jsbundling-rails
+
+First ensure that you have Yarn and Node installed. Run ```yarn -v``` and ```node -v``` in the terminal. Both of these should return a number indicating the version. 
+
+
+<span id="bundler-check"></span>Now let's go through these steps with a new app and for the last step select esbuild for your JavaScript. Run the command below.</span>
+~~~bash
+rails new testapp
+~~~
+Add the jsbundling-rails gem to your Gemfile with.
+
+~~~ruby
+gem 'jsbundling-rails'
+~~~
+
+Then run the command below in the terminal within your Rails application folder.
+
+~~~bash 
+./bin/bundle install
+~~~
+
+Remember the three options for the bundler are esbuild, rollup and webpack and you pick only one. 
+
+~~~bash
+./bin/rails javascript:install:<replace the text and <> with your bundler choice>
+~~~
+
+You should have now see this error and a few other ones but don't fear!
+
+~~~bash
+✘ [ERROR] Could not resolve "@hotwired/turbo-rails"
+
+    app/javascript/application.js:2:7:
+      2 │ import "@hotwired/turbo-rails"
+        ╵        ~~~~~~~~~~~~~~~~~~~~~~~
+~~~ 
+
+It's quite explicit in the issue. Resolve is just a fancy word for saying the bundler wasn't able to do anything with this import. Which makes sense as our testapp was set up initially with import maps instead of with esbuild or another bundler. Now let's see if we can resolve this in our ```app/javascript/application.js```. It looks fine so what gives? Remember that this Rails app was started with _import maps_ not a bundler. We need to install the package so that way it can be bundled appropriately. Now let's run the command below.
+
+~~~bash
+npm i @hotwired/turbo-rails
```

Comment:
Why do we use `npm` here rather than yarn?

Reply:
Mainly because that was the command used by dhh [here](https://m.youtube.com/watch?v=Chiu-0EVW3g&t=639s). More than happy although to change it to Yarn!

Reply:
I think using npm, if yarn is already there, will cause a problem with competing package lock versions. Let's just stick to yarn since that is what we have people install.

Reply:
Done!